### PR TITLE
Rework our system header approach to be more error proof

### PR DESCRIPTION
**Pull request by miscco**

### CODE REVIEW

**Review by wmaxey — APPROVED** [review]

**Review by griwes — APPROVED** [review]

**Review by elstehle — APPROVED** [review]

**Review by gevtushenko — APPROVED** [review]

File: libcudacxx/include/cuda/std/detail/libcxx/include/__cccl_config [hunk 1]
```diff
@@ -22,6 +22,10 @@
 #  define _CCCL_COMPILER_MSVC
 #elif defined(__IBMCPP__)
 #  define _CCCL_COMPILER_IBM
+#elif defined(__INTEL_LLVM_COMPILER)
```
Comment (miscco) [inline libcudacxx/include/cuda/std/detail/libcxx/include/__cccl_config:25]: Funnily enough we were also missing intel compiler in our list

File: libcudacxx/include/cuda/std/detail/libcxx/include/__cccl_config [hunk 2]
```diff
@@ -47,46 +51,44 @@
 #endif
 
 // Enforce that cccl headers are treated as system headers
-#if defined(_CCCL_COMPILER_GCC)
-#  define _CCCL_FORCE_SYSTEM_HEADER _Pragma("GCC system_header")
+#if defined(_CCCL_COMPILER_GCC) || defined(_CCCL_COMPILER_NVHPC) || defined(_CCCL_COMPILER_ICC) \
```
Comment (miscco) [inline libcudacxx/include/cuda/std/detail/libcxx/include/__cccl_config:54]: This is the real change. We filter compilers and use the appropriate #pragma depending on the define

File: libcudacxx/include/cuda/std/detail/libcxx/include/experimental/propagate_const [hunk 3]
```diff
@@ -106,11 +106,13 @@
 
 */
 
-#include <experimental/__config>#if defined(_CCCL_COMPILER_NVHPC) && defined(_CCCL_USE_IMPLICIT_SYSTEM_DEADER)
-#pragma GCC system_header
-#else // ^^^ _CCCL_COMPILER_NVHPC ^^^ / vvv !_CCCL_COMPILER_NVHPC vvv
-_CCCL_IMPLICIT_SYSTEM_HEADER
-#endif // !_CCCL_COMPILER_NVHPC
+#include <experimental/__config>#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
```
Comment (gevtushenko) [inline libcudacxx/include/cuda/std/detail/libcxx/include/experimental/propagate_const:109]: important: 
```suggestion
#include <experimental/__config>

#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
```

**Review by gevtushenko — APPROVED** [review]

**Comment by jrhemstad** [comment]
@miscco will manually backport this PR by cherry-picking https://github.com/NVIDIA/cccl/commit/774a2a552e9c42a9e5085e9aef2baa835daa2f69

**Pull request merged**